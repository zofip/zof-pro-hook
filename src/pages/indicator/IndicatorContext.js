import { createContext } from 'react';

const IndicatorContext = createContext(null);

export default IndicatorContext;