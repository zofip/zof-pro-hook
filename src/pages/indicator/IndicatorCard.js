import React, { useState, useContext } from 'react';

import clsx from 'clsx';
import { makeStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardContent from '@material-ui/core/CardContent';
import CardActions from '@material-ui/core/CardActions';
import Collapse from '@material-ui/core/Collapse';
import Avatar from '@material-ui/core/Avatar';
import IconButton from '@material-ui/core/IconButton';
import Typography from '@material-ui/core/Typography';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import AddIcon from '@material-ui/icons/Add';
import RemoveIcon from '@material-ui/icons/Remove';
import ViewModule from '@material-ui/icons/ViewModule';
import CheckCircle from '@material-ui/icons/CheckCircle';

import IndicatorContext from './IndicatorContext';
import ValidateInput from './ValidateInput';
import { ADD, REMOVE, PERCENT, EMPTY } from '../../utils/constants';


const useStyles = makeStyles(theme => ({
    card: {
        maxWidth: 345,
    },
    media: {
        height: 0,
        paddingTop: '56.25%', // 16:9
    },
    expand: {
        transform: 'rotate(0deg)',
        transition: theme.transitions.create('transform', {
            duration: theme.transitions.duration.shortest,
        }),
    },
    expandOpen: {
        transform: 'rotate(180deg)',
    },
    avatar: {
        color: '#009BB0',
        background: '#D9D6D0'
    },
    add: {
        marginLeft: 'auto'
    }
}));


const IndicatorCard = ({ indicator }) => {

    const dispatch = useContext(IndicatorContext);

    const classes = useStyles();
    const [expanded, setExpanded] = useState(false);
    const [add, setAdd] = useState(null);
    const [value, setValue] = useState();

    const handleInputValueChange = newValue => {
        if (newValue.target) {
            setValue(newValue.target.value);
        } else {
            setValue(newValue);
        }
    }

    const handleExpandClick = () => {
        setExpanded(!expanded);
    }

    const handleAddClick = () => {
        if (value) {
            if (indicator.metric.unit === PERCENT) {
                setValue(40);
                setAdd(true);
                dispatch({
                    type: ADD,
                    indicator: indicator,
                    value: 40
                });
            } else {
                setAdd(true);
                dispatch({
                    type: ADD,
                    indicator: indicator,
                    value: value
                });
            }
        }
    }

    const handleRemoveClick = () => {
        setAdd(false);
        dispatch({
            type: REMOVE,
            event: indicator.metric.name
        });
    }

    return (
        <Card className={classes.card}>
            <CardHeader
                avatar={
                    <Avatar aria-label="recipe" className={add ? classes.avatar : EMPTY}>
                        {add ? <CheckCircle /> : <ViewModule />}
                    </Avatar>
                }
                title={indicator.metric.name}
                subheader={indicator.metric.type}
            />
            <CardContent>
                <Typography variant="body2" color="textSecondary" component="p">
                    kpi: {indicator.kpi}
                </Typography>
                <ValidateInput unit={indicator.metric.unit} handleInputValueChange={handleInputValueChange} />
            </CardContent>
            <CardActions disableSpacing>
                <IconButton
                    className={clsx(classes.expand, {
                        [classes.expandOpen]: expanded,
                    })}
                    onClick={handleExpandClick}
                    aria-expanded={expanded}
                    aria-label="show more">
                    <ExpandMoreIcon />
                </IconButton>
                {add === false || add === null ?
                    <IconButton className={classes.add} aria-label="add" color="secondary" onClick={handleAddClick}>
                        <AddIcon />
                    </IconButton>
                    :
                    <IconButton className={classes.add} aria-label="remove" color="default" onClick={handleRemoveClick}>
                        <RemoveIcon />
                    </IconButton>
                }
            </CardActions>
            <Collapse in={expanded} timeout="auto" unmountOnExit>
                <CardContent>
                    <Typography paragraph>Description:</Typography>
                    <Typography variant="body2" color="textSecondary">{indicator.metric.description}</Typography>
                </CardContent>
            </Collapse>
        </Card>
    );
}

export default IndicatorCard;