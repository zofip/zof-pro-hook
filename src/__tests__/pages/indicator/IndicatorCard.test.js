import React from 'react';
import { shallow } from 'enzyme';

import addedIndicatorsReducer from '../../../pages/indicator/reducers/addedIndicatorReducer';
import IndicatorContext from '../../../pages/indicator/IndicatorContext';
import IndicatorCard from '../../../pages/indicator/IndicatorCard';
import { jsxEmptyExpression } from '@babel/types';

describe('<IndicatorCard />', () => {
   it('Rendering', () => {    
      
      const indicatorPercent = {
        'componentId': 1,
        'extractorId': 3,
        'graphic': "Areas",
        'id': 3,
        'kpi': ">65",
        'metric': {
            'description': "Cobertura de código a través de pruebas unitarias. (Obtener de Sonar) ",
            'id': 3,
            'name': "coverage",
            'type': "Técnico / Código",
            'unit': "PORCENTAJE"
        },
        'metricId': 3,
        'percintil_alert': 40
      };  

      const indicatorNumber = {
         'componentId': 1,
         'extractorId': 12,
         'graphic': "Barras",
         'id': 9,
         'kpi': "<40",
         'metric': {
            'description': "Un problema relacionado con la mantenibilidad en el código. Dejarlo como está significa que, en el mejor de los casos, los mantenedores tendrán más dificultades de lo que deberían hacer cambios en el código. En el peor de los casos, estarán tan confundidos por el estado del código que introducirán errores adicionales a medida que realicen cambios. (Obtiene de Sonar).",
            'id': 4,
            'name': "code_smell",
            'type': "Técnico / Código",
            'unit': "CANTIDAD"
         },
         'metricId': 4,
         'percintil_alert': 50
       }; 

      const TestComponent = () => (
      <IndicatorContext.Provider value={jest.fn()} >
            <IndicatorCard indicator={indicatorPercent} />
      </IndicatorContext.Provider>
     );

      const indicatorCardPercent = shallow(<TestComponent />);
      expect(indicatorCardPercent).toMatchSnapshot();
      const indicatorCardNumber = shallow(<IndicatorCard indicator={indicatorNumber} />);
      expect(indicatorCardNumber).toMatchSnapshot();
   });
});