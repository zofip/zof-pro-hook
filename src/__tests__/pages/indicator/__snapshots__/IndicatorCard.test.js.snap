// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IndicatorCard /> Rendering 1`] = `
<ContextProvider
  value={[MockFunction]}
>
  <IndicatorCard
    indicator={
      Object {
        "componentId": 1,
        "extractorId": 3,
        "graphic": "Areas",
        "id": 3,
        "kpi": ">65",
        "metric": Object {
          "description": "Cobertura de código a través de pruebas unitarias. (Obtener de Sonar) ",
          "id": 3,
          "name": "coverage",
          "type": "Técnico / Código",
          "unit": "PORCENTAJE",
        },
        "metricId": 3,
        "percintil_alert": 40,
      }
    }
  />
</ContextProvider>
`;

exports[`<IndicatorCard /> Rendering 2`] = `
<WithStyles(ForwardRef(Card))
  className="makeStyles-card-1"
>
  <WithStyles(ForwardRef(CardHeader))
    avatar={
      <ForwardRef(WithStyles)
        aria-label="recipe"
        className=""
      >
        <Memo />
      </ForwardRef(WithStyles)>
    }
    subheader="Técnico / Código"
    title="code_smell"
  />
  <WithStyles(ForwardRef(CardContent))>
    <WithStyles(ForwardRef(Typography))
      color="textSecondary"
      component="p"
      variant="body2"
    >
      kpi: 
      &lt;40
    </WithStyles(ForwardRef(Typography))>
    <ValidateInput
      handleInputValueChange={[Function]}
      unit="CANTIDAD"
    />
  </WithStyles(ForwardRef(CardContent))>
  <WithStyles(ForwardRef(CardActions))
    disableSpacing={true}
  >
    <WithStyles(ForwardRef(IconButton))
      aria-expanded={false}
      aria-label="show more"
      className="makeStyles-expand-3"
      onClick={[Function]}
    >
      <ExpandMoreIcon />
    </WithStyles(ForwardRef(IconButton))>
    <WithStyles(ForwardRef(IconButton))
      aria-label="add"
      className="makeStyles-add-6"
      color="secondary"
      onClick={[Function]}
    >
      <AddIcon />
    </WithStyles(ForwardRef(IconButton))>
  </WithStyles(ForwardRef(CardActions))>
  <WithStyles(ForwardRef(Collapse))
    in={false}
    timeout="auto"
    unmountOnExit={true}
  >
    <WithStyles(ForwardRef(CardContent))>
      <WithStyles(ForwardRef(Typography))
        paragraph={true}
      >
        Description:
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        color="textSecondary"
        variant="body2"
      >
        Un problema relacionado con la mantenibilidad en el código. Dejarlo como está significa que, en el mejor de los casos, los mantenedores tendrán más dificultades de lo que deberían hacer cambios en el código. En el peor de los casos, estarán tan confundidos por el estado del código que introducirán errores adicionales a medida que realicen cambios. (Obtiene de Sonar).
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(CardContent))>
  </WithStyles(ForwardRef(Collapse))>
</WithStyles(ForwardRef(Card))>
`;
